docs/website/contents/for-developers/HardWonWisdom.md

```diff
@@ -348,3 +348,165 @@ The key corresponding parts of the implementation are as follows.
+      The safe zone cannot be 0 because---as discussed else where---that would imply a stability window of 0, which is impossible.
+- An instance of `CanHardFork [F, G]`, which decomposes as follows.
+    - A function `LedgerConfig F -> LedgerConfig G -> LedgerState F -> Bound {- starting slot&epoch&time-since-genesis of G -} -> SlotNo -> Except OutsideForecastRange (LedgerView G)`.
+      This function will be used for forecasting the ledger view across the era transition; see "How does cross-era forecasting work?" above.
+    - A function `ConsensusConfig F -> ConsensusConfig G -> EpochNo {- start of G -} -> ChainDepState F -> ChainDepState G`.
+      This function will be applied to the chain-dependent state immediately after the last `F` header.
+      Then the protocol rules for `G` will be used to tick that state to the slot of the first `G` header.
+      This use of `G`'s ticking rules instead of `F`'s to cross the end of the last epoch of `F` is subtle and somewhat counter-intuitive.
+      (We haven't yet had any concrete issues here, unlike with the `LedgerState` issue linked just below.
+      However, we did only realize recently that the extra entropy parameter of Transitional Praos should have been considered when translating to Praos at the Alonzo->Babbage transition.)
+    - A function `LedgerConfig F -> LedgerConfig G -> EpochNo {- start of G -} -> LedgerState F -> LedgerState G`.
+      This function will be applied to the ledger state immediately after the last `F` block.
+      Then the ledger rules for `G` will be used to tick that state to the slot of the first `G` block.
+      This use of `G`'s ticking rules instead of `F`'s to cross the end of the last epoch of `F` is subtle and somewhat counter-intuitive (eg see Issue <https://github.com/input-output-hk/cardano-ledger/issues/3491>).
+    - Instances of `SingleEraBlock` for `F` and `G`, which decomposes as the following for `F` and `G` individually.
+        - Types that represent the ledger and protocol configuration _without_ any `EpochInfo` (see `HasPartialLedgerConfig` and `HasPartialConsensusConfig`).
+        - A function `singleEraTransition :: PartialLedgerConfig blk -> EraParams -> Bound {- era's start -} -> LedgerState blk -> Maybe EpochNo` which must return `Just e` if and only if the arguments determine that only a roll back of more than `k` blocks (which is impossible without disk corruption or manual intervention) could prevent this era from ending between epoch `e` and its predecessor (ie epoch `e-1` when `0<e` and "genesis" otherwise).
+            - As an immediate observation: note that this function is monotonic along the ledger states of an era on a single chain.
+              It cannot return `Just` for some ledger state on a chain and then `Nothing` for a later ledger state in the same era on the same chain.
```

Comment:
I'm not certain that the Consensus Layer _inherently_ requires this property, and so I hesitate to include it here. Thanks for mentioning it, though.

Reply:
We briefly talked about this yesterday, and we should then probably also remove/tweak the mention of rollbacks just above:
> A function `singleEraTransition :: PartialLedgerConfig blk -> EraParams -> Bound {- era's start -} -> LedgerState blk -> Maybe EpochNo` which must return `Just e` if and only if the arguments determine that only a roll back of more than `k` blocks (which is impossible without disk corruption or manual intervention) could prevent this era from ending between epoch `e` and its predecessor (ie epoch `e-1` when `0<e` and "genesis" otherwise).

Reply:
Fixed.